runtime/src/active_stats.rs

```diff
@@ -0,0 +1,64 @@
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+#[derive(Debug, Default)]
+pub struct ActiveStats {
+    clean: AtomicUsize,
+    shrink: AtomicUsize,
+    hash: AtomicUsize,
+    flush: AtomicUsize,
+}
+
+#[derive(Debug, Copy, Clone)]
+pub enum ActiveStatItem {
+    Clean,
+    Shrink,
+    Hash,
+    Flush,
+}
+
+/// sole purpose is to handle 'drop' so that stat is decremented when self is dropped
+pub struct ActiveState<'a> {
+    stats: &'a ActiveStats,
+    item: ActiveStatItem,
+}
+
+impl<'a> Drop for ActiveState<'a> {
+    fn drop(&mut self) {
+        self.stats.log(self.item, false);
+    }
+}
+
+impl ActiveStats {
+    #[must_use]
+    /// create a stack object to set the state to increment stat initially and decrement on drop
+    pub fn get_state(&self, stat: ActiveStatItem) -> ActiveState<'_> {
+        self.log(stat, true);
+        ActiveState {
+            stats: self,
+            item: stat,
+        }
+    }
+    /// update and log the change to the specified 'item'
+    fn log(&self, item: ActiveStatItem, increment: bool) {
+        let stat = match item {
+            ActiveStatItem::Clean => &self.clean,
+            ActiveStatItem::Shrink => &self.shrink,
+            ActiveStatItem::Hash => &self.hash,
+            ActiveStatItem::Flush => &self.flush,
+        };
+        if increment {
+            stat.fetch_add(1, Ordering::Relaxed);
+        } else {
+            stat.fetch_sub(1, Ordering::Relaxed);
+        }
```

Comment:
I know you've been using this pattern a lot, but in this case we always know when we're incrementing (creating the ActiveState) and when we're decrementing (on drop). If `log` took an update fn, then creating the ActiveState could be:

```rust
self.log(stat, |stat| stat.fetch_add(1, Ordering::Relaxed));
```

and drop could be:
```rust
self.stats.log(self.item, |stat| stat.fetch_sub(1, Ordering::Relaxed));
```

And then this block would collapse to 
```rust
update_fn(stat);
```